### ISSUE TITLE
Incorrect number of tricks when allowing duplicates in the combo generator

### ISSUE THREAD

**Issue by JulianDietzel** [bastislack/highline-freestyle#226]
## Issue

When generating a combo and allowing duplicate tricks, it can happen that in the result contains less tricks than requested by the user. 

## Steps to reproduce

This issue does not always occur. Even the same settings for the generator can lead to both faulty and correct results when rerunning the generation process.

I strongly suspect that it issue arises from duplicates being dropped. So one way that has not failed to produce the issue for me is by requiring more tricks in the combo than there are in the specified difficulty range. This guarantees the generation of duplicates  (which then don't show up). Therefore one exemplary way to reproduce the issue is this:

1. Navigate to the Combo Generator
2. Set the number of tricks to 20
3. Set the minimum and maximum difficulty both to 1
4. Click on "Advanced"
5. Check the "Allow duplicate tricks" checkbox
6. Hit the "Generate" Button

## Additional details

The "Number of tricks" shown in the "Combo Stats" section is in line with the number of tricks requested by the user but does not match the shown number of tricks when the issue occurs.

**Comment by JulianDietzel** [comment]
After further testing, the source of the issue has now become more clear:

## Overview
The generator function (`findCombo`) works properly. When generating a combo and directly printing the generated combo to the console after the function returns, the output contains the correct number of tricks in a sensible order.

The problem occurs in the `ComboDetails` which is invoked by `ComboGenerator`. Here, the generated combo is [passed into it](https://github.com/bastislack/highline-freestyle/blob/9e5ff8ef30f1fc6f74ab154cbf69c4acb9c24b57/src/components/combos/ComboDetails.js#L18) as the `comboToShow` variable. This variable (/object) contains all the information that is needed to display the combo to my eyes.

Then the [following line](https://github.com/bastislack/highline-freestyle/blob/9e5ff8ef30f1fc6f74ab154cbf69c4acb9c24b57/src/components/combos/ComboDetails.js#L41) is executed in the `ComboDetails`:
```js
const combo = useLiveQuery(() => queryFunc(), [comboToShow]);
```
This creates the variable `combo` which is then used to display the combo to the user. It is, verifiably, the `queryFunc` function, which causes the issue. Not only does it remove duplicate tricks but it also reorders all tricks in the combo by their id.


## Analysis of `queryFunc`

The function is split into an if- and an else-block:

[Option a)](https://github.com/bastislack/highline-freestyle/blob/9e5ff8ef30f1fc6f74ab154cbf69c4acb9c24b57/src/components/combos/ComboDetails.js#L31) is that the combo to be displayed is being generated or posted. This leads to the existing combo's tricks being stripped of all their information except their ids and then `db.fillComboWithTricks(comboToShow)`  being called which adds all the removed info back to the tricks.

[Option b)](https://github.com/bastislack/highline-freestyle/blob/9e5ff8ef30f1fc6f74ab154cbf69c4acb9c24b57/src/components/combos/ComboDetails.js#L35) is that the combo already exists. Then it is fetched via `db.getCombo(params.id)`.

Since `db.getCombo()` internally also calls `db.fillComboWithTricks()` which then calls `db.getTricksByIds`, this last function is the one to investigate. And it turns out to be "the root" of the issue as the [first line of it](https://github.com/bastislack/highline-freestyle/blob/9e5ff8ef30f1fc6f74ab154cbf69c4acb9c24b57/src/services/db.js#L171) reads:
```js
ids = ids.map(id => Number(id));
```
The passed ids are mapped. Which naturally removes duplicates and, incidentally, also orders them in ascending order. After this line is called, the original order and all duplicates are gone.


## Solutions

I see one necessary and one optional solution to this. 
1. **Necessary:** Change the `db.getTricksByIds()` function so that duplicates and the original order are preserved. This should be without any side effects, as `db.fillComboWithTricks()`is the only function calling it.
2. **Optional:** Remove the database request entirely in the `queryFunc` function, at least when a generated combo is passed. The data is already there, removing and then readding it seems a little strange.


## Questions

This leads me to my main question: Why is this operation of removing the data and then adding it back again performed. There clearly must have been intention in doing it, as it happens within the span of two adjacent lines of code.
I changed the line
```js
const combo = useLiveQuery(() => queryFunc(), [comboToShow]);
```
to
```js
const combo = comboToShow
```
and it works (only for when using the generator of course, this is _not_ the fix, just a test).

So why is this performed and even more importantly, is there potential for anything breaking if this call to the database is removed. I am not very familiar with React and want to make sure that I am not missing anything (especially since the comments in the code mention that this relates to React Hooks) before I make a pull request.

Sorry for the wall of text.

Git Blame tells me that one of you guys might be able to help me with this: @weberax  @bastislack

**Comment by weberax** [comment]
> Why is this operation of removing the data and then adding it back again performed.

This was done because we had the problem that the Tricks showed with the ComboDetail Component where not live updated, i.e. when changing the Stickfrequency of one of the Tricks of a Combo, this was not visible (the color) when navigating back.  (see #115)

**Comment by JulianDietzel** [comment]
Fixed in #227
